### PR TITLE
[distributed][Tensor Parallelism] Implement early transforms for column-wise and row-wise linear and embedding

### PR DESCRIPTION
**Pull request by crcrpar**
this implements a trace transform that converts one or more linear and/or embedding layers into column-wise or row-wise tensor parallel ones by (1) sharding their weight and bias and (2) inserting needed communication and/or scattering before and/or after the modified layers.

Out of four supported ops, row-wise parallel linear would lead to a `BoundSymbol` modification. The change caused is to omit the bias term and that bias is added to the result of communication (after post-processing). 

---

example

```python
class Model(nn.Module):
    def __init__(self, n_in: int, n_hidden: int, n_out: int) -> None:
        self.l1 = nn.Linear(n_in, n_hidden)
        self.l2 = nn.Linear(n_hidden, n_out)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.l2(F.gelu(self.l1(x))


device = torch.device(f"cuda:{rank}")

model = Model().to(device)
jitted_model = thunder.jit(model)
tp_jitted_model = thunder.distributed.column_parallel(jitted_model, ("l1",))
tp_jitted_model = thunder.distributed.row_parallel(tp_jitted_model, ("l2",))

x = torch.randn(..., device=device)
y = tp_jitted_model(x)
assert y.size(1) == n_out
```

cc @borda @apaz-cli @carmocca @awaelchli @crcrpar

### CODE REVIEW

**Comment by crcrpar** [comment]
The failures as of https://github.com/Lightning-AI/lightning-thunder/commit/f724a886639bc616c93879435efcbbeb4c8ac2fe look related to https://github.com/Lightning-AI/lightning-thunder/issues/432.

File: thunder/distributed/tensor_parallel.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+            visit=visit,
+            provenance="gather ouptut of column-wise tensor parallel layer",
+        )
+        if distributed_c10d.get_rank() == 0:
+            print(visit.swap_map)
+
+        return prologue_trace, new_computation_trace, epilogue_trace
+
+
+# TODO(crcrpar): Support `nn.Embedding`
```
Comment (crcrpar) [inline thunder/distributed/tensor_parallel.py:224]:
```suggestion
```

File: thunder/distributed/tensor_parallel.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+# TODO(crcrpar): Add an option to turn off output all-gather.
```
Comment (crcrpar) [inline thunder/distributed/tensor_parallel.py:225]: feels like this kind of optimizations should come after this early transform as another early transform

File: thunder/distributed/tensor_parallel/row_wise.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        self,
+        computation_trace: TraceCtx,
+    ) -> tuple[Callable[[BoundSymbol], VISIT_TYPE], TraceProvenance | str]:
+        from thunder.core.pytree import tree_flatten
+
+        consumers = utils.consumers(computation_trace)
+        flat_args, _ = tree_flatten((computation_trace.args, computation_trace.kwargs))
+        bsym2prepostprocess: dict[BoundSymbol, PrePostProcessInterface] = {}
+        for proxy in filter(lambda p: isinstance(p, TensorProxy), flat_args):
+            for p_name in self.chunked_param_name2layer_type:
```
Comment (lantiga) [inline thunder/distributed/tensor_parallel/row_wise.py:117]: We could simplify and remove the inner for loop replace it with

```python
    layer_type = self.chunked_param_name2layer_type.get(proxy.name)
    if layer_type is not None:
        ...
        if ...:
            match layer_type:
                case ...
```

File: thunder/executors/torchex.py [hunk 4]
```diff
@@ -1755,8 +1755,16 @@ def _all_gather_prim_impl(
         /,
         group: torch.distributed.ProcessGroup,
         do_async: Number,
+        dim: int | None = None,
     ) -> torch.Tensor | tuple[torch.distributed.distributed_c10d.Work, torch.Tensor]:
-        out: torch.Tensor = torch.empty((group.size() * a.shape[0],) + a.shape[1:], dtype=a.dtype, device=a.device)
+        result_shape = list(a.shape)
+        if dim is not None:
+            utils.check_type(dim, int)
+            utils.check(dim >= 0 and dim < a.dim(), lambda: f"dim must satisfry 0 <= {dim=} < {a.dim()=}")
```
Comment (lantiga) [inline thunder/executors/torchex.py:1763]:
```suggestion
            utils.check(dim >= 0 and dim < a.dim(), lambda: f"dim must satisfy 0 <= {dim=} < {a.dim()=}")
```

File: thunder/tests/distributed/test_ddp.py [hunk 5]
```diff
@@ -1021,6 +1037,169 @@ def fwd_loss(m, x):
...
 
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="")
+    @common_utils.parametrize("name", ("column", "row"))
+    def test_tensor_parallel_linear(self, name):
+        device = torch.device("cuda", self.rank)
+        x = torch.randn(2, 12).to(device).requires_grad_()
+
+        process_group = None
+        ref_model = ToyModel().to(device)
+        with c10d._coalescing_manager(async_ops=True) as cm:
```
Comment (lantiga) [inline thunder/tests/distributed/test_ddp.py:1048]: for my own edification: why are we all_reducing here?

File: thunder/tests/distributed/test_ddp.py [hunk 6]
```diff
@@ -1021,6 +1037,169 @@ def fwd_loss(m, x):
...
+            c10d.all_reduce(x)
+            for p in ref_model.parameters():
+                c10d.all_reduce(p)
+        cm.wait()
+
+        ref_state_dict = ref_model.state_dict()
+        expected = ref_model(x)
+
+        converter = {
```
Comment (lantiga) [inline thunder/tests/distributed/test_ddp.py:1057]: `converter` -> `transform`?

File: thunder/tests/distributed/test_ddp.py [hunk 7]
```diff
@@ -1021,6 +1037,169 @@ def fwd_loss(m, x):
...
+        ref_state_dict = ref_model.state_dict()
+        expected = ref_model(x)
+
+        model = Model().to(device)
+        model.load_state_dict(ref_state_dict)
+        jitted_model = thunder.jit(model)
+
+        column_parallels = ["embed_1", "linear1_0", "linear2_1"]
+        row_parallels = ["embed_2", "linear1_1", "linear2_0"]
+        tp_jitted_model = convert_module_to_rowwise_parallel(
```
Comment (lantiga) [inline thunder/tests/distributed/test_ddp.py:1184]: just a comment: in the future it would be great to have high-level logic to dispatch row parallel and column parallel based on how ops on the trace are laid out.

**Review by lantiga — APPROVED** [review]
Amazing work @crcrpar 

Mostly nitpicks, it was great fun reviewing this

File: thunder/distributed/tensor_parallel/column_wise.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+    def get_visitor_of_computation_trc_and_provenance(
+        self,
+        computation_trace: TraceCtx,
+    ) -> tuple[Callable[[BoundSymbol], VISIT_TYPE], TraceProvenance | str]:
+        from thunder.core.pytree import tree_flatten
+
+        consumers = utils.consumers(computation_trace)
+        flat_args, _ = tree_flatten((computation_trace.args, computation_trace.kwargs))
+        bsym2prepostprocess: dict[BoundSymbol, PrePostProcessInterface] = {}
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/distributed/tensor_parallel/column_wise.py:117]: (+ the same change below)

File: thunder/distributed/tensor_parallel/column_wise.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+            world_size = int(os.environ["WORLD_SIZE"])
+            local_rank = int(os.environ["LOCAL_RANK"])
+            device = torch.device(f"cuda:{local_rank}")
+            distributed_c10d.init_process_group()
+            model = Model().to(device)
+            jitted_model = thunder.jit(model)
+            # `l2`'s output size (= n_out) needs to be divisible by `world_size`
+            tp_model = convert_module_to_columnwise_parallel(
+                jitted_model,
+                target_modules=("embed", "l2",),
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/distributed/tensor_parallel/column_wise.py:202]: but if you use this example, I think `transformer.h.0.mlp.fc_1`

File: thunder/distributed/tensor_parallel/column_wise.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+    # Modify module
+    for module_name, _ in thunder_module._model.named_modules():
+        if module_name not in target_modules:
+            continue
+        submodule = thunder_module.get_submodule(module_name)
+
+        for pn, p in submodule.named_parameters(recurse=False, prefix=module_name):
+            # if we don't have an override or it is just the original, do create a copy
+            if thunder_module._overrides.get(pn, p) is p:
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/distributed/tensor_parallel/column_wise.py:243]: This needs to be `_overrides_parameters`, but we should split out the common bits between here and fsdp.

**Review by t-vi — APPROVED** [review]
Supergood, LGTM. Very excited to see the `test_tensor_parallel_both_column_and_row` be the first actual example of composing `early_transforms`!
I added a few minor nits.

File: thunder/distributed/prims.py [hunk 11]
```diff
@@ -270,6 +289,41 @@ def stash_grad_for_fsdp_meta(
     return TensorProxy(like=grad)
 
 
+# see [Megatron-LM: Training Multi-Billion Parameter Language Models Using Model Parallelism](https://arxiv.org/abs/1909.08053)'s Code 1.
+def synchronize_input_for_column_wise_tensor_parallel_meta(
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply thunder/distributed/prims.py:293]: nope, this is removed in https://github.com/crcrpar/lightning-thunder/tree/crpa/comm-optimize_tensor-parallel which is based on this diff. I'd have to clean this up

File: thunder/distributed/prims.py [hunk 12]
```diff
@@ -330,3 +389,74 @@ def synchronize_backward_rule(
...
+        case TensorParallelLayerType.COLUMN_PARALLEL_LINEAR:
+            return (
+                reduce_scatter(
+                    grad / group.size(), DistributedReduceOps.SUM, group, do_async=True, dim=grad.ndim - 1
+                ).wait(),
+                None,
+                None,
+            )
+        case TensorParallelLayerType.ROW_PARALLEL_LINEAR:
+            return grad, None, None
```
Comment (t-vi) [inline thunder/distributed/prims.py:450]: Thinking about it more: I know nothing about it, but if AllReduce produces 4 copies of the output, should the backward sum the for gradients corresponding to it? (based on the picture as https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allreduce )

File: thunder/distributed/tensor_parallel/row_wise.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        chunk_size = x.shape[x.ndim - 1] // self.process_group.size()
+        start_idx = chunk_size * c10d.get_rank(self.process_group)
+        return clang.slice_in_dim(x, start_idx, start_idx + chunk_size, dim=x.ndim - 1), None
+
+    def postprocess(self, y: TensorProxy, _: Any) -> TensorProxy:
+        # gather `y` along the last dimension
+        import thunder.torch as ltorch
+        from thunder.distributed import prims as dist_prims
+
+        all_reduced = dist_prims.synchronize_tensor_parallel_output(
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply thunder/distributed/tensor_parallel/row_wise.py:55]: adding a prim for each `layer_type` felt lackluster to me; writing a meta, creating a symbol, writing forward rule and backward rule. also I thought it's similar to what `dist_prims.synchronize` does accordingly to `t.ddp_type`

File: thunder/tests/distributed/test_ddp.py (hunk 5 above)
Comment (crcrpar) [reply thunder/tests/distributed/test_ddp.py:1048]: all the ranks have the same initial weights, which I don't think mandatory for the purpose

File: thunder/tests/distributed/test_ddp.py [hunk 14]
```diff
@@ -1021,6 +1037,169 @@ def fwd_loss(m, x):
         with thunder.ThunderModule.no_sync(model):
             fwd_loss(model, x)
 
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="")
+    @common_utils.parametrize("name", ("column", "row"))
+    def test_tensor_parallel_linear(self, name):
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply thunder/tests/distributed/test_ddp.py:1042]: no objections but we should've done way earlier...

File: thunder/distributed/__init__.py [hunk 15]
```diff
@@ -589,33 +593,40 @@ def _shard_param(
...
-                ),
-            )
-        chunk_size = param.shape[0] // world_size
-        # NOTE This could be a ShardTensor to indicate other parts of the code
-        # that it's sharded and should be treated differently
-        shard = param.data.narrow(0, chunk_size * rank, chunk_size).clone()
-        param.data = shard
-    else:
+    dim_to_shard = 0 if dim is None else dim
+    if allow_padding_for_fsdp:
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply thunder/distributed/__init__.py:602]: would there be needs for padding by distributed parallel schemes other than fsdp?
I embarrassingly am not aware of any

File: thunder/distributed/__init__.py [hunk 16]
```diff
@@ -27,6 +29,8 @@
     "fsdp",
     "FSDPBucketingStrategy",
     "FSDPType",
+    "convert_module_to_columnwise_parallel",
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply thunder/distributed/__init__.py]: going for shorter names

File: thunder/distributed/prims.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (crcrpar) [reply thunder/distributed/prims.py]: done in https://github.com/Lightning-AI/lightning-thunder/pull/410/commits/0eb89fc5a14786ccd26a9d30f811cb73813646b9

File: thunder/distributed/tensor_parallel/row_wise.py (hunk 3 above)
Comment (crcrpar) [reply thunder/distributed/tensor_parallel/row_wise.py]: done in e2a8a08274d9e185401fd5c9844c0208eb6b085d

File: thunder/tests/distributed/test_tensor_parallel.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+        y.mean().backward()
+
+        torch.testing.assert_close(
+            expected=ref_model.embed.weight.grad.chunk(self.world_size, dim)[self.rank],
+            actual=tp_jitted_model.get_parameter("embed.weight").grad,
+        )
+
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="")
+    @common_utils.parametrize("bias", (True, False))
+    def test_tensor_parallel_both_column_and_row(self, bias):
```
Comment (crcrpar) [inline thunder/tests/distributed/test_tensor_parallel.py:140]: one more fix is needed. the grads don't match

File: thunder/tests/distributed/test_tensor_parallel.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+        for col, orig_size in zip(column_parallel_layers, [num_embeddings, n_hidden]):
+            weight = tp_model.get_parameter(f"{col}.weight")
+            self.assertEqual(weight.size(0), orig_size // self.world_size)
+        for row, orig_size in zip(row_parallel_layers, [embedding_dim, n_hidden]):
+            weight = tp_model.get_parameter(f"{row}.weight")
+            self.assertEqual(weight.size(1), orig_size // self.world_size)
+
+        torch.testing.assert_close(actual=actual, expected=expected)
+        torch.testing.assert_close(actual=x.grad, expected=x_ref.grad)
```
Comment (crcrpar) [inline thunder/tests/distributed/test_tensor_parallel.py:188]: tensors of long type, meaningless
```suggestion
```

File: thunder/core/proxies.py [hunk 19]
```diff
@@ -994,10 +994,13 @@ def __repr__(self):
         return f"[FloatProxy name={self.name}, value={self.value}]"
 
 
-class DDPType(Enum):
+class DistParallelType(Enum):
```
Comment (t-vi) [inline thunder/core/proxies.py:997]: I like the name change. :)

File: thunder/tests/distributed/test_tensor_parallel.py (hunk 17 above)
Comment (t-vi) [reply thunder/tests/distributed/test_tensor_parallel.py:144]: Great job finding and fixing it!

File: thunder/distributed/prims.py [hunk 20]
```diff
@@ -330,3 +389,74 @@ def synchronize_backward_rule(
         case _:
             utils.check(False, lambda: f"synchronize with unexpected {ddp_type=}")
     return synced_grad, None
+
+
+@register_augmented_forward(PrimIDs.SYNCHRONIZE_TENSOR_PARALLEL_OUTPUT)
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply thunder/distributed/prims.py:456]: Both are equivalent in functionality. The future is to allow either way of specifying the reverse differentiation.

File: thunder/distributed/tensor_parallel/column_wise.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/distributed/tensor_parallel/column_wise.py:213]: yep, my bad

File: thunder/distributed/tensor_parallel/common.py [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+__all__ = [
+    "ComputationTraceTransformVisitorForTensorParallel",
+    "TensorParallelLayerType",
+    "NoOp",
+    "PrePostProcessInterface",
+    "TransformForTensorParallel",
+]
+
+
+class TensorParallelLayerType(Enum):
```
[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/distributed/tensor_parallel/common.py:35]: Right, my point is that PyTorch doesn't have the "layer" terminology. So it should probably be `TensorParallelModuleType(Enum)`

File: thunder/distributed/tensor_parallel/row_wise.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/distributed/tensor_parallel/row_wise.py:58]: sounds good, we can always revisit later

**Review by lantiga — APPROVED** [review]
Looks great! Ship it! 🚀

File: thunder/executors/torchex.py [hunk 22]
```diff
@@ -1755,8 +1755,16 @@ def _all_gather_prim_impl(
         /,
         group: torch.distributed.ProcessGroup,
         do_async: Number,
+        dim: int | None = None,
     ) -> torch.Tensor | tuple[torch.distributed.distributed_c10d.Work, torch.Tensor]:
-        out: torch.Tensor = torch.empty((group.size() * a.shape[0],) + a.shape[1:], dtype=a.dtype, device=a.device)
+        result_shape = list(a.shape)
+        if dim is not None:
```
[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/executors/torchex.py:1812]: I was also thinking about this, we can leave this for the future

File: thunder/distributed/tensor_parallel/common.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/distributed/tensor_parallel/common.py:35]: We can take this on later
